### PR TITLE
Controlnet guidance for f3d texture stage

### PR DESCRIPTION
**Pull request by yankeesong**
For the texture stage, used normal map generated by the fixed mesh for fine grained control from controlnet. 

An example prompt:
"a girl with undercut blue hair and cowboy boots"

https://github.com/threestudio-project/threestudio/assets/36912641/4bdc8cc1-0db2-4158-884d-923094ebf3af

with usual SDS guidance, the hair completely covers the ears.

https://github.com/threestudio-project/threestudio/assets/36912641/abbad81b-46ce-44c9-9a5a-e723765a723f

with controlnet guidance, the ears are much more visible.

### CODE REVIEW

File: threestudio/systems/fantasia3d.py [hunk 1]
```diff
@@ -68,9 +68,15 @@ def training_step(self, batch, batch_idx):
             )
         else:  # texture training
             guidance_inp = out["comp_rgb"]
-            guidance_out = self.guidance(
-                guidance_inp, prompt_utils, **batch, rgb_as_latents=False
-            )
+            cond_inp = out["comp_normal"]
```
Comment (bennyguo) [inline threestudio/systems/fantasia3d.py:71]: How about moving this line under the `if`?

**Review by bennyguo — COMMENTED** [review]
Look great!

File: threestudio/systems/fantasia3d.py (hunk 1 above)
Comment (yankeesong) [reply threestudio/systems/fantasia3d.py]: Done!

File: threestudio/systems/fantasia3d.py [hunk 2]
```diff
@@ -68,9 +68,15 @@ def training_step(self, batch, batch_idx):
...
+            if "controlnet" in self.cfg.guidance_type:
```
[the remark this answers is not in the document]
Comment (yankeesong) [reply threestudio/systems/fantasia3d.py]: Done!

**Review by bennyguo — APPROVED** [review]

**Comment by dunbar12138** [comment]
Hey, thanks for your contribution! A quick question: are you aware of #189? Controlnet seems to take view-space normals while I believe `comp_normal` is in the world space.

Let me know if I miss anything!

**Comment by yankeesong** [comment]
True. Thanks for pointing it out! I'll make the modifications.

**Pull request merged**